ts/features/services/components/ServiceDetailsPreferences.tsx

```diff
@@ -0,0 +1,162 @@
+import React, { ComponentProps, useCallback, useEffect } from "react";
+import { FlatList, ListRenderItemInfo } from "react-native";
+import {
+  Divider,
+  IOToast,
+  ListItemHeader,
+  ListItemSwitch
+} from "@pagopa/io-app-design-system";
+import * as O from "fp-ts/lib/Option";
+import * as RA from "fp-ts/lib/ReadonlyArray";
+import { pipe } from "fp-ts/lib/function";
+import { NotificationChannelEnum } from "../../../../definitions/backend/NotificationChannel";
+import { ServiceId } from "../../../../definitions/backend/ServiceId";
+import I18n from "../../../i18n";
+import { useIODispatch, useIOSelector } from "../../../store/hooks";
+import { isPremiumMessagesOptInOutEnabledSelector } from "../../../store/reducers/backendStatus";
+import { EnabledChannels } from "../../../utils/profile";
+import { useFirstEffect } from "../common/hooks/useFirstEffect";
+import { upsertServicePreference } from "../store/actions";
+import {
+  isErrorServicePreferenceSelector,
+  isLoadingServicePreferenceSelector,
+  servicePreferenceResponseSuccessSelector
+} from "../store/reducers/servicePreference";
+import { serviceMetadataInfoSelector } from "../store/reducers/servicesById";
+
+const hasChannel = (
+  notificationChannel: NotificationChannelEnum,
+  channels: ReadonlyArray<NotificationChannelEnum> = []
+) =>
+  pipe(
+    channels,
+    RA.findFirst(channel => channel === notificationChannel),
+    O.isSome
+  );
+
+type PreferenceSwitchListItem = {
+  condition?: boolean;
+} & ComponentProps<typeof ListItemSwitch>;
+
+export type ServiceDetailsPreferencesProps = {
+  serviceId: ServiceId;
+  availableChannels?: ReadonlyArray<NotificationChannelEnum>;
+};
+
+export const ServiceDetailsPreferences = ({
+  serviceId,
+  availableChannels = []
+}: ServiceDetailsPreferencesProps) => {
+  const isFirstRender = useFirstEffect();
```

Comment:
using this custom hook seems a bit of a stretch, could there be other ways? For example you could use the usePrevius hook on isErrorServicePreference... when prev is undefined it means you are in the first render
What do you think?

Reply:
In this case `usePrevious` doesn't solve our case:

- if we enter for the first time in the service details, `isErrorServicePreference` is `false` and the fetch fails so the toast is displayed correctly.
- If we go back to the service list and re-enter in the same service details, then `isErrorServicePreference` is `true` and the fetch fails so two toasts are displayed.

so, if we want to use the `usePrevious` hook, we should clear the `ServicePreferenceState` when go back to the service list, instead with the pr solution we avoid this step. What do you think?

Reply:
Yes, so, it's perfect! 🔝